forge/src/executor/inspector/cheatcodes/env.rs

```diff
@@ -0,0 +1,171 @@
+use std::collections::BTreeMap;
+
+use super::Cheatcodes;
+use crate::abi::HEVMCalls;
+use bytes::Bytes;
+use ethers::{
+    abi::{AbiEncode, Token, Tokenize},
+    types::{Address, H256, U256},
+    utils::keccak256,
+};
+use revm::{Database, EVMData};
+
+#[derive(Clone, Debug, Default)]
+pub struct Prank {
+    /// Address of the contract that initiated the prank
+    pub prank_caller: Address,
+    /// Address of `tx.origin` when the prank was initiated
+    pub prank_origin: Address,
+    /// The address to assign to `msg.sender`
+    pub new_caller: Address,
+    /// The address to assign to `tx.origin`
+    pub new_origin: Option<Address>,
+    /// The depth at which the prank was called
+    pub depth: u64,
+    /// Whether or not the prank stops by itself after the next call
+    pub single_call: bool,
+}
+
+fn prank(
+    state: &mut Cheatcodes,
+    prank_caller: Address,
+    prank_origin: Address,
+    new_caller: Address,
+    new_origin: Option<Address>,
+    depth: u64,
+    single_call: bool,
+) -> Result<Bytes, Bytes> {
+    let prank = Prank {
+        prank_caller,
+        prank_origin,
+        new_caller,
+        new_origin,
+        // Note: When calling a cheatcode, the depth is not increased. Because of that, the depth
+        // we actually want to prank at is 1 level deeper.
+        depth: depth + 1,
+        single_call,
+    };
+
+    if state.prank.is_some() {
+        return Err("You have an active prank already.".to_string().encode().into())
+    }
+
+    state.prank = Some(prank);
+    Ok(Bytes::new())
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct RecordAccess {
+    pub reads: BTreeMap<Address, Vec<U256>>,
+    pub writes: BTreeMap<Address, Vec<U256>>,
+}
+
+fn record(state: &mut Cheatcodes) {
```

Comment:
Is this right? Shouldn't it be taking some extra argument to record? Else, maybe should be called "start record" or something?

Reply:
Takes no arguments, but we can rename to start record:

https://github.com/gakonst/foundry/blob/a00c1939becf3b40f1453e330d3d9507a756d610/evm-adapters/src/sputnik/cheatcodes/cheatcode_handler.rs#L812-L815